src/zope/interface/adapter.py

```diff
@@ -303,13 +456,61 @@ def unsubscribe(self, required, provided, value=None):
                 del byorder[-1]
 
         if provided is not None:
-            n = self._provided[provided] + len(new) - len(old)
+            n = self._provided[provided] + len(new) - len_old
             if n == 0:
                 del self._provided[provided]
                 self._v_lookup.remove_extendor(provided)
+            else:
+                self._provided[provided] = n
 
         self.changed(self)
 
+    def rebuild(self):
+        """
+        Rebuild (and replace) all the internal data structures of this
+        object.
+
+        This is useful, especially for persistent implementations, if
+        you suspect an issue with reference counts keeping interfaces
+        alive even though they are no longer used.
+
+        It is also useful if you or a subclass change the data types
+        (``_mappingType`` and friends) that are to be used.
+
+        This method replaces all internal data structures with new objects;
+        it specifically does not re-use any storage.
+
+        .. versionadded:: 5.3.0
+        """
+
+        # Grab the iterators, we're about to discard their data.
+        registrations = self.allRegistrations()
+        subscriptions = self.allSubscriptions()
+
+        def buffer(it):
+            # The generator doesn't actually start running until we
+            # ask for its next(), by which time the attributes will change
+            # unless we do so before calling __init__.
+            from itertools import chain
```

Comment:
should this import go to the top?

Reply:
I was trying to avoid importing it unless it was needed, since this function will rarely be called. But I hadn't actually checked to see if that made a difference at all.

It turns out that `import pkg_resources` (which happens in `zope/__init__.py`) imports `itertools` always. Even when that can go away (when/if we go to pure Python-3-only namespace packages), somewhere in the startup process of Python 3.6 and above `itertools` is imported before user code even executes.

So it's essentially a free import and can be put at module scope with no penalty.